src/posix/platform/multicast_routing.cpp

```diff
@@ -0,0 +1,560 @@
+#include <netinet/in.h>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <unistd.h>
+#if __linux__
+#include <linux/mroute6.h>
+#else
+#error "Multicast Routing feature is not ported to non-Linux platforms yet."
+#endif
+
+#include <openthread/backbone_router_ftd.h>
+
+#include "core/common/logging.hpp"
+
+namespace ot {
+namespace Posix {
+
+void MulticastRoutingManager::Init(otInstance *aInstance)
+{
+    otBackboneRouterSetMulticastListenerCallback(aInstance,
+                                                 &MulticastRoutingManager::HandleBackboneMulticastListenerEvent, this);
+}
+
+void MulticastRoutingManager::HandleBackboneMulticastListenerEvent(void *                                 aContext,
+                                                                   otBackboneRouterMulticastListenerEvent aEvent,
+                                                                   const otIp6Address *                   aAddress)
+{
+    static_cast<MulticastRoutingManager *>(aContext)->HandleBackboneMulticastListenerEvent(
+        aEvent, static_cast<const Ip6::Address &>(*aAddress));
+}
+
+void MulticastRoutingManager::HandleBackboneMulticastListenerEvent(otBackboneRouterMulticastListenerEvent aEvent,
+                                                                   const Ip6::Address &                   aAddress)
+{
+    switch (aEvent)
+    {
+    case OT_BACKBONE_ROUTER_MULTICAST_LISTENER_ADDED:
+        Add(aAddress);
+        break;
+    case OT_BACKBONE_ROUTER_MULTICAST_LISTENER_REMOVED:
+        Remove(aAddress);
+        break;
+    }
+}
+
+void MulticastRoutingManager::Enable(void)
+{
+    otError error = InitMulticastRouterSock();
+
+    otLogResultPlat(error, "MulticastRoutingManager: %s", __FUNCTION__);
+}
+
+void MulticastRoutingManager::Disable(void)
+{
+    FinalizeMulticastRouterSock();
+
+    otLogResultPlat(OT_ERROR_NONE, "MulticastRoutingManager: %s", __FUNCTION__);
+}
+
+void MulticastRoutingManager::Add(const Ip6::Address &aAddress)
+{
+    assert(mListenerSet.find(aAddress) == mListenerSet.end());
+    mListenerSet.insert(aAddress);
+
+    VerifyOrExit(IsEnabled());
+
+    UnblockInboundMulticastForwardingCache(aAddress);
+
+exit:
+    otLogResultPlat(OT_ERROR_NONE, "MulticastRoutingManager: %s: %s", __FUNCTION__, aAddress.ToString().AsCString());
+}
+
+void MulticastRoutingManager::Remove(const Ip6::Address &aAddress)
+{
+    otError error = OT_ERROR_NONE;
+
+    assert(mListenerSet.find(aAddress) != mListenerSet.end());
+    mListenerSet.erase(aAddress);
+
+    VerifyOrExit(IsEnabled());
+
+    RemoveInboundMulticastForwardingCache(aAddress);
+
+exit:
+    otLogResultPlat(error, "MulticastRoutingManager: %s: %s", __FUNCTION__, aAddress.ToString().AsCString());
+}
+
+void MulticastRoutingManager::UpdateFdSet(fd_set &aReadFdSet, int &aMaxFd) const
+{
+    VerifyOrExit(IsEnabled());
+
+    FD_SET(mMulticastRouterSock, &aReadFdSet);
+    aMaxFd = std::max(aMaxFd, mMulticastRouterSock);
```

Comment:
avoid std::max since we intentionally removed dependency on stdc++ in #3130.

Reply:
Is it necessary for POSIX?

Reply:
changed to use OT_MAX